integration/benchmark_test.go

```diff
@@ -0,0 +1,239 @@
+// +build performance
+
+package integration
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	_ "github.com/aquasecurity/fanal/analyzer/command/apk"
+	_ "github.com/aquasecurity/fanal/analyzer/library/bundler"
+	_ "github.com/aquasecurity/fanal/analyzer/library/cargo"
+	_ "github.com/aquasecurity/fanal/analyzer/library/composer"
+	_ "github.com/aquasecurity/fanal/analyzer/library/npm"
+	_ "github.com/aquasecurity/fanal/analyzer/library/pipenv"
+	_ "github.com/aquasecurity/fanal/analyzer/library/poetry"
+	_ "github.com/aquasecurity/fanal/analyzer/library/yarn"
+	_ "github.com/aquasecurity/fanal/analyzer/os/alpine"
+	_ "github.com/aquasecurity/fanal/analyzer/os/amazonlinux"
+	_ "github.com/aquasecurity/fanal/analyzer/os/debianbase"
+	_ "github.com/aquasecurity/fanal/analyzer/os/photon"
+	_ "github.com/aquasecurity/fanal/analyzer/os/redhatbase"
+	_ "github.com/aquasecurity/fanal/analyzer/os/suse"
+	_ "github.com/aquasecurity/fanal/analyzer/pkg/apk"
+	_ "github.com/aquasecurity/fanal/analyzer/pkg/dpkg"
+	_ "github.com/aquasecurity/fanal/analyzer/pkg/rpm"
+
+	"github.com/aquasecurity/fanal/analyzer"
+	"github.com/aquasecurity/fanal/cache"
+	"github.com/aquasecurity/fanal/extractor/docker"
+	"github.com/aquasecurity/fanal/types"
+	godeptypes "github.com/aquasecurity/go-dep-parser/pkg/types"
+	dtypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type testCase struct {
+	name                 string
+	imageName            string
+	imageFile            string
+	expectedFiles        []string
+	expectedOS           analyzer.OS
+	expectedPkgsFromCmds string
+	expectedLibraries    string
+}
+
+func runChecksBench(j int, b *testing.B, ac analyzer.Config, ctx context.Context, tc testCase, d string, c cache.Cache) {
+	for i := 0; i < b.N; i++ {
+		var wg sync.WaitGroup
+		for i := 0; i <= j; i++ {
+			wg.Add(1)
+			go func(i int) {
+				defer wg.Done()
+
+				actualFiles, err := ac.Analyze(ctx, tc.imageFile)
```

Comment:
`Analyze` takes the most time. So, I think the same thing as mentioned above happens.

Commit A
- Analyze: 1000ms
- GetOS: 10ms
- GetPackages: 10ms
- GetLibraries: 10ms
- Total: 1030ms

Commit B
- Analyze: 1000ms
- GetOS: 50ms
- GetPackages: 10ms
- GetLibraries: 10ms
- Total: 1080ms

I think we should split those functions into each benchmark.

Reply:
I personally feel this is nice to have but is a bit of an overkill for benchmarking. What you're really after here is tracing I feel. Knowing a breakdown by function/dependency is better off being done with APM/Tracing than measuring individual benchmarks as the former is more meaningful.

It would also make our tests excruciatingly long to read and brittle to maintain.

Reply:
It is possible with distributed tracing. OpenTracing is one option. It does require additional work to set it up (collection of traces, etc.) But it's something we can embed into a distributable like Trivy CLI and gather metrics. https://github.com/opentracing/opentracing-go